src/NuGet.Core/NuGet.Packaging/Signing/Verification/AllowListVerificationProvider.cs

```diff
@@ -36,35 +34,66 @@ private PackageVerificationResult VerifyAllowList(ISignedPackageReader package,
             if (_allowList.Count() > 0 && !IsSignatureAllowed(signature))
             {
                 status = SignatureVerificationStatus.Invalid;
-                issues.Add(SignatureLog.Issue(fatal: true, code: NuGetLogCode.NU3003, message: string.Format(CultureInfo.CurrentCulture, Strings.Error_NoMatchingCertificate, _fingerprintAlgorithm.ToString())));
+                issues.Add(SignatureLog.Error(NuGetLogCode.NU3003, Strings.Error_NoAllowedCertificate));
             }
 
             return new SignedPackageVerificationResult(status, signature, issues);
         }
 
         private bool IsSignatureAllowed(PrimarySignature signature)
         {
-            // Get information needed for allow list verification
-            var primarySignatureCertificateFingerprint = CertificateUtility.GetHash(signature.SignerInfo.Certificate, _fingerprintAlgorithm);
-            var primarySignatureCertificateFingerprintString = BitConverter.ToString(primarySignatureCertificateFingerprint).Replace("-", "");
+            // To avoid getting the countersignature more than once lets memoize it
+            var repositoryCounterSignature = new Lazy<RepositoryCountersignature>(() => RepositoryCountersignature.GetRepositoryCountersignature(signature));
+
+            // Also memoize certificate fingerprints for primary signature and repository countersignature
+            var signatureCertFingerprints = new Dictionary<HashAlgorithmName, string>();
+            var countersignatureCertFingerprints = new Dictionary<HashAlgorithmName, string>();
 
             foreach (var allowedEntry in _allowList)
             {
                 // Verify the certificate hash allow list objects
-                var certificateHashEntry = allowedEntry as CertificateHashAllowListEntry;
-                if (certificateHashEntry != null)
+                var certHashEntry = allowedEntry as CertificateHashAllowListEntry;
+                if (certHashEntry != null)
                 {
-                    if (certificateHashEntry.VerificationTarget.HasFlag(VerificationTarget.Primary) &&
-                        StringComparer.OrdinalIgnoreCase.Equals(certificateHashEntry.CertificateFingerprint, primarySignatureCertificateFingerprintString))
+                    if (certHashEntry.VerificationTarget.HasFlag(VerificationTarget.Primary))
                     {
-                        return true;
+                        if (!signatureCertFingerprints.TryGetValue(certHashEntry.FingerprintAlgorithm, out var signatureCertFingerprint))
+                        {
+                            signatureCertFingerprint = GetCertFingerprint(signature.SignerInfo.Certificate, certHashEntry.FingerprintAlgorithm);
```

Comment:
is it intentional that you are immediately overwriting the value of your out param?

Reply:
if `signatureCertFingerprints.TryGetValue(certHashEntry.FingerprintAlgorithm, out var signatureCertFingerprint)` does not succeed it will go inside the if statement and `signatureCertFingerprint` will be `null` so I set it first thing so I can use it outside. Then I add it to the hashmap so I can use it the next time without calculating it again.

Reply:
aah i see. didn't notice the ``` ! ``` :(